File: App/SelfReportFlow.swift
// 템포루틴 — 앱 내 자기보고 설문 화면 (v1.6 §4 / MASTER §3.11)
//
// 웹 설문과 다른 점 하나: **결과를 보여주지 않는다.**
// 앱은 이미 기록에서 유형을 계산하는데(§5.12), 자기보고 유형까지 화면에 띄우면
// 산출 경로가 다른 두 값이 나란히 보이게 된다. 어느 쪽이 "내 유형"인지 물어보는 순간
// 답이 없다. 그래서 응답은 저장만 하고 대조는 내부에서만 쓴다.
//
// 마무리 카피도 이득을 약속하지 않는다 — 자기보고를 실제로 쓰는 경로가 아직 없어서,
// "더 정확해져요" 류는 검증 불가능한 주장이 된다(§7 가드레일).

import SwiftUI
import SwiftData
import TempoCore

struct SelfReportFlow: View {
    @Environment(\.modelContext) private var modelContext
    @Environment(\.dismiss) private var dismiss

    @State private var step = 0
    @State private var answers: [String: String] = [:]
    /// 제시 순서만 섞는다. 한 번 섞은 순서를 유지해야 화면을 오갈 때 문항이 튀지 않는다.
    @State private var symptomOrder: [SurveyQuestion] = SelfReportSurvey.symptomQuestions.shuffled()

    private var totalSteps: Int { 5 }
    /// 스크롤 리셋 대상(2026-08-12) — 장이 바뀌면 여기로 되돌린다.
    private static let topAnchor = "survey-top"

    /// 문항이 하나뿐인 장 — 선택이 곧 그 장의 답이라 「다음」을 한 번 더 누를 이유가 없다
    /// (2026-08-12 베타 피드백 "선택하면 다음 안눌러도 넘어가게"). 1장 = 캘리브레이션 단문항.
    private var isSingleQuestionStep: Bool { step == 1 }

    var body: some View {
        NavigationStack {
            ZStack {
                Ink.paper.ignoresSafeArea()
                // "설문지" 인상 걷어내기(2026-08-08 베타 피드백 "하기 싫게 생겼어") —
                // 온보딩과 같은 겨울 광 + 감쇠 텍스처로 앱의 지면 어휘를 잇는다
                SeasonLight(phase: .menstrual, motif: .onboarding)
                // 장을 넘기면 맨 위부터 보여준다(2026-08-12 베타 피드백) — 종전엔 이전 장의
                // 스크롤 위치가 남아 새 장이 중간부터 열렸다.
                ScrollViewReader { proxy in
                    ScrollView {
                        VStack(alignment: .leading, spacing: 18) {
                            progressHeader.id(Self.topAnchor)
                            content
                            actions
                        }
                        .padding(20)
                        .centeredColumn(640)
                    }
                    .onChange(of: step) { _, _ in
                        proxy.scrollTo(Self.topAnchor, anchor: .top)
                    }
                }
            }
            .navigationTitle("리듬 설문")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("닫기") { dismiss() }.foregroundStyle(Ink.text)
                }
            }
        }
    }

    // 진행도 명시(2026-08-08 베타 피드백 "진행도도 표시해줘") — 얇은 실선만으론 안 읽혔다
    private var progressHeader: some View {
        HStack(spacing: 10) {
            GeometryReader { proxy in
                ZStack(alignment: .leading) {
                    Capsule().fill(Ink.text.opacity(0.12))
                    Capsule().fill(Ink.text.opacity(0.65))
                        .frame(width: max(step > 0 ? 6 : 0,
                                          proxy.size.width * Double(step) / Double(totalSteps)))
                }
            }
            .frame(height: 4)
            if step > 0 {
                Text("\(step) / \(totalSteps)")
                    .font(.caption)
                    .monospacedDigit()
                    .foregroundStyle(Ink.text.opacity(0.55))
            }
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(Loc.fmt("진행도 %1$@ / %2$@", "\(step)", "\(totalSteps)"))
    }

    @ViewBuilder
    private var content: some View {
        switch step {
        case 0: intro
        case 1: questionGroup(title: nil,
                              note: "이제 그때를 떠올리면서 답해주세요.",
                              questions: [SelfReportSurvey.calibration])
        case 2: questionGroup(title: nil, note: nil, questions: SelfReportSurvey.phaseQuestions)
        case 3: questionGroup(title: symptomAnchorTitle(p2: answers["P2"]),
                              note: "꼭 생리 중이 아니라도 가장 힘들었던 때를 떠올려주세요.",
                              questions: symptomOrder)
        case 4: questionGroup(title: nil, note: nil, questions: SelfReportSurvey.amplitudeQuestions)
        default: optionalGroup
        }
    }

    /// 증상 문항의 시간 앵커 문구 — TempoCore의 `symptomAnchorLine`은 선택지 라벨을 문장에
    /// **합성**해서 돌려주므로 그대로는 번역 키가 아니다. 순수 모듈을 건드리지 않으려고
    /// 합성만 앱으로 가져온다(판정 규칙·선택지 자체는 그대로 TempoCore가 소유).
    private func symptomAnchorTitle(p2: String?) -> String {
        guard let p2, p2 != "none", p2 != "unknown",
              let choice = SelfReportSurvey.phaseChoices.first(where: { $0.value == p2 })
        else { return String(localized: "그나마 힘들었던 때를 떠올려서, 평소와 비교해서 답해주세요.") }
        return Loc.fmt("「%1$@」, 평소와 비교해서 답해주세요.", Loc.text(choice.label))
    }

    private var intro: some View {
        VStack(alignment: .leading, spacing: 12) {
            // 줄바꿈 = 의미 단위로(2026-08-05 베타 피드백 "줄바꿈 신경써줘")
            Text("이맘때 이야기를 하려면\n몇 가지 알아야 해요.")
                .font(.almanac(size: 26, weight: .bold))
                .foregroundStyle(Ink.text)
            Text("리듬의 모양은 사람마다 달라요.\n2분 정도 소요돼요.")
                .font(.system(.body, design: .serif))
                .foregroundStyle(Ink.text.opacity(0.75))
            Text("응답은 이 기기에만 저장됩니다.")
                .font(.footnote)
                .foregroundStyle(Ink.text.opacity(0.5))
        }
    }

    private func questionGroup(title: String?, note: String?, questions: [SurveyQuestion]) -> some View {
        VStack(alignment: .leading, spacing: 22) {
            if let title {
                Text(title)
                    .font(.almanac(size: 19, weight: .bold))
                    .foregroundStyle(Ink.text)
            }
            if let note {
                Text(note).font(.footnote).foregroundStyle(Ink.text.opacity(0.55))
            }
            ForEach(questions, id: \.id) { question in
                questionBlock(question)
            }
        }
    }

    private var optionalGroup: some View {
        VStack(alignment: .leading, spacing: 22) {
            VStack(alignment: .leading, spacing: 4) {
                Text("이 페이지의 질문은 건너뛰셔도 괜찮아요.")
                    .font(.footnote).foregroundStyle(Ink.text.opacity(0.55))
            }
            ForEach(SelfReportSurvey.optionalQuestions, id: \.id) { question in
                questionBlock(question)
            }
        }
    }

    // 문항 = 세리프 표제 + 라디오·괘선 리스트(2026-08-08 베타 피드백 — 회색 캡슐 12개가
    // 벽처럼 쌓이던 조판 폐기, 책력 개방 조판으로). 측정 로직·저장 키는 무변경.
    private func questionBlock(_ question: SurveyQuestion) -> some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(question.text)
                .font(.almanacBody(.subheadline, size: 16, weight: .bold))
                .foregroundStyle(Ink.text)
                .padding(.bottom, 4)
            ForEach(question.choices, id: \.value) { choice in
                choiceRow(question: question, choice: choice)
            }
        }
    }

    private func choiceRow(question: SurveyQuestion, choice: SurveyChoice) -> some View {
        let selected = answers[question.id] == choice.value
        return Button {
            select(question: question, choice: choice, wasSelected: selected)
        } label: {
            HStack(spacing: 10) {
                Image(systemName: selected ? "circle.inset.filled" : "circle")
                    .font(.subheadline)
                    .foregroundStyle(selected ? Ink.text : Ink.text.opacity(0.3))
                Text(choice.label)
                    .font(.subheadline.weight(selected ? .semibold : .regular))
                    .foregroundStyle(Ink.text.opacity(selected ? 1.0 : 0.75))
                Spacer(minLength: 0)
            }
            .padding(.vertical, 11)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .almanacRule(opacity: 0.12)
        .accessibilityAddTraits(selected ? [.isSelected] : [])
    }

    // 2026-08-05 베타 피드백: 다음 = 우하단 글씨만(검정 캡슐 제거), 이전 = 모든 문항 단계에.
    private var actions: some View {
        HStack(spacing: 18) {
            if step > 0 {
                Button("이전") { step -= 1 }
                    .font(.body.weight(.medium))
                    .foregroundStyle(Ink.text.opacity(0.55))
            }
            Spacer(minLength: 0)
            if step == totalSteps {
                Button("건너뛰기") { finish() }
                    .font(.body)
                    .foregroundStyle(Ink.text.opacity(0.55))
            }
            Button(step == totalSteps ? "제출" : "다음") { advance() }
                .font(.body.weight(.semibold))
                .foregroundStyle(canAdvance ? Ink.text : Ink.text.opacity(0.3))
                .disabled(!canAdvance)
        }
        .padding(.top, 6)
    }

    /// 선택 문항 단계 말고는 그 화면의 문항이 전부 채워져야 넘어간다.
    private var canAdvance: Bool {
        switch step {
        case 0: true
        case 1: answers[SelfReportSurvey.calibration.id] != nil
        case 2: SelfReportSurvey.phaseQuestions.allSatisfy { answers[$0.id] != nil }
        case 3: SelfReportSurvey.symptomQuestions.allSatisfy { answers[$0.id] != nil }
        case 4: SelfReportSurvey.amplitudeQuestions.allSatisfy { answers[$0.id] != nil }
        default: true
        }
    }

    /// 선택 반영 + 단문항 장이면 자동 진행(2026-08-12). 재탭 해제로는 넘어가지 않는다.
    /// 한 박자 늦추는 이유 = 고른 표시(라디오 채움)를 보고 넘어가야 무엇을 골랐는지 남는다.
    private func select(question: SurveyQuestion, choice: SurveyChoice, wasSelected: Bool) {
        answers[question.id] = wasSelected ? nil : choice.value
        guard !wasSelected, isSingleQuestionStep else { return }
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 260_000_000)
            // 사이에 사용자가 직접 넘겼거나 답을 지웠으면 손대지 않는다
            guard isSingleQuestionStep, canAdvance else { return }
            advance()
        }
    }

    private func advance() {
        if step < totalSteps {
            step += 1
        } else {
            finish()
        }
    }

    private func finish() {
        // 화이트리스트 밖 키가 섞이지 않게 한 번 거른다(웹 서버와 같은 규칙)
        let allowed = SelfReportSurvey.allQuestionIDs
        let cleaned = answers.filter { allowed.contains($0.key) }
        modelContext.insert(SelfReportRecord(answers: cleaned))
        SelfReportStore.hasPrompted = true
        dismiss()
    }
}

File: android/app/src/main/kotlin/app/temporoutine/android/selfreport/SelfReportFlow.kt
// 템포루틴 Android — 앱 내 자기보고 설문 (iOS SelfReportFlow.swift 이식, v1.6 §4 / MASTER §3.11)
// 결과를 보여주지 않는다 — 앱은 기록에서 유형을 계산하는데(§5.12) 자기보고 유형까지 띄우면 산출 경로가 다른 두 값이 나란히 보인다.
// 문항 순서·선택지·판정은 tempocore SelfReportSurvey가 소유(증상 8문항만 제시 순서 셔플, 시간 앵커 = P2 응답).
// 단문항 장(1장)은 선택 즉시 자동 진행하되 최초 1회만(답 고치러 돌아온 사람을 다시 밀어내지 않는다, 2026-09-04 베타).

package app.temporoutine.android.selfreport

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import app.temporoutine.android.R
import app.temporoutine.android.onboarding.SurveyLogic
import app.temporoutine.android.theme.Fonts
import app.temporoutine.android.theme.Ink
import app.temporoutine.android.theme.MotifStyle
import app.temporoutine.android.theme.SeasonLight
import app.temporoutine.core.CyclePhase
import app.temporoutine.core.SelfReportSurvey
import app.temporoutine.core.SurveyChoice
import app.temporoutine.core.SurveyQuestion
import dev.chrisbanes.haze.hazeSource
import dev.chrisbanes.haze.rememberHazeState
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * @param previousAnswers 다시 하기 — 이전 응답을 채운 채 연다(결과 화면이 없는 앱이라 이전 응답을 「보는」 유일한 길).
 * @param onSubmit 화이트리스트 필터 전 원본 응답. 저장은 호출자(VM)가 한다.
 * @param onDismiss 제출·건너뛰기·닫기 어느 경로든 시트가 걷힌 뒤 1회.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SelfReportFlow(previousAnswers: Map<String, String> = emptyMap(), onSubmit: (Map<String, String>) -> Unit, onDismiss: () -> Unit) {
    val ink = Ink
    val scope = rememberCoroutineScope()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val hazeState = rememberHazeState()
    val scroll = rememberScrollState()

    var step by rememberSaveable { mutableIntStateOf(0) }
    val answers = remember { mutableStateMapOf<String, String>().apply { putAll(previousAnswers) } }
    val prefilled = previousAnswers.isNotEmpty()
    /** 제시 순서만 섞는다. 한 번 섞은 순서를 유지해야 화면을 오갈 때 문항이 튀지 않는다. */
    val symptomOrder = remember { SelfReportSurvey.symptomQuestions.shuffled() }
    /** 자동 진행을 이미 태운 장 — 답을 고치러 돌아온 사람을 다시 밀어내면 고칠 방법이 없다. */
    var autoAdvanced by remember { mutableStateOf(setOf<Int>()) }
    /** 자동 진행 예약(장 번호) — 260ms 뒤 소비. 사이에 사용자가 직접 넘겼거나 답을 지웠으면 손대지 않는다. */
    var armedStep by remember { mutableStateOf<Int?>(null) }
    val canAdvance = SurveyLogic.canAdvance(step, answers)

    fun close() { scope.launch { sheetState.hide() }.invokeOnCompletion { onDismiss() } }
    fun finish() { onSubmit(answers.toMap()); close() }
    fun advance() { if (step < SurveyLogic.TOTAL_STEPS) step += 1 else finish() }
    /** 선택 반영 + 단문항 장이면 자동 진행 예약. 재탭 해제로는 넘어가지 않는다. */
    fun select(question: SurveyQuestion, choice: SurveyChoice) {
        val wasSelected = choice.value in SelfReportSurvey.values(answers[question.id])
        if (question.allowsMultiple) {
            val next = SurveyLogic.toggledMultiValue(question, choice, answers[question.id])
            if (next == null) answers.remove(question.id) else answers[question.id] = next
        } else if (wasSelected) {
            answers.remove(question.id)
        } else {
            answers[question.id] = choice.value
        }
        if (!wasSelected && SurveyLogic.isSingleQuestionStep(step) && step !in autoAdvanced) armedStep = step
    }

    // 장을 넘기면 맨 위부터(2026-08-12 베타 피드백)
    LaunchedEffect(step) { scroll.scrollTo(0) }
    // 한 박자 늦추는 이유 = 고른 표시(라디오 채움)를 보고 넘어가야 무엇을 골랐는지 남는다
    LaunchedEffect(armedStep) {
        val armed = armedStep ?: return@LaunchedEffect
        delay(260)
        armedStep = null
        if (step != armed || !SurveyLogic.isSingleQuestionStep(step) || !SurveyLogic.canAdvance(step, answers)) return@LaunchedEffect
        autoAdvanced = autoAdvanced + armed
        advance()
    }

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState, containerColor = ink.paper, dragHandle = null) {
        Box(Modifier.fillMaxSize()) {
            // "설문지" 인상 걷어내기 — 온보딩과 같은 겨울 광 + 감쇠 텍스처로 지면 어휘를 잇는다
            SeasonLight(phase = CyclePhase.MENSTRUAL, modifier = Modifier.fillMaxSize().hazeSource(hazeState), motif = MotifStyle.ONBOARDING)
            Column(Modifier.fillMaxSize()) {
                Box(Modifier.fillMaxWidth().padding(horizontal = 8.dp, vertical = 6.dp)) {
                    TextButton(onClick = ::close, modifier = Modifier.align(Alignment.CenterStart)) {
                        Text(stringResource(R.string.survey_close), style = Fonts.system(17), color = ink.text)
                    }
                    Text(stringResource(R.string.survey_title), style = Fonts.system(17, FontWeight.SemiBold), color = ink.text, modifier = Modifier.align(Alignment.Center))
                }
                Column(
                    Modifier
                        .fillMaxWidth()
                        .widthIn(max = 640.dp)
                        .align(Alignment.CenterHorizontally)
                        .verticalScroll(scroll)
                        .padding(20.dp),
                    verticalArrangement = Arrangement.spacedBy(18.dp),
                ) {
                    ProgressHeader(step)
                    when (step) {
                        0 -> Intro(prefilled)
                        1 -> QuestionGroup(null, listOf(stringResource(R.string.survey_note_calibration)), listOf(SelfReportSurvey.calibration), answers, ::select)
                        // 복수 응답 안내(2026-09-09) — 체크 모양만으로는 여러 개를 골라도 된다는 걸 모른다
                        2 -> QuestionGroup(null, listOf(stringResource(R.string.survey_note_multiple)), SelfReportSurvey.phaseQuestions, answers, ::select)
                        3 -> QuestionGroup(symptomAnchorTitle(answers["P2"]), listOf(stringResource(R.string.survey_note_symptom)), symptomOrder, answers, ::select)
                        4 -> QuestionGroup(null, emptyList(), SelfReportSurvey.amplitudeQuestions, answers, ::select)
                        // 저장 위치 재고지(2026-09-09 베타) — 인트로에 이미 한 줄 있지만, 답하기를 망설이는 자리는 이 페이지다
                        else -> QuestionGroup(
                            null,
                            listOf(stringResource(R.string.survey_note_optional), stringResource(R.string.survey_note_optional_local)),
                            SelfReportSurvey.optionalQuestions, answers, ::select,
                        )
                    }
                    // 다음 = 우하단 글씨만, 이전 = 모든 문항 단계에(2026-08-05 베타 피드백)
                    Row(Modifier.fillMaxWidth().padding(top = 6.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(18.dp)) {
                        if (step > 0) ActionText(stringResource(R.string.survey_prev), enabled = true, weight = FontWeight.Medium, alpha = 0.55f) { step -= 1 }
                        Spacer(Modifier.weight(1f))
                        if (step == SurveyLogic.TOTAL_STEPS) ActionText(stringResource(R.string.survey_skip), enabled = true, weight = FontWeight.Normal, alpha = 0.55f) { finish() }
                        ActionText(
                            if (step == SurveyLogic.TOTAL_STEPS) stringResource(R.string.survey_submit) else stringResource(R.string.ob_next),
                            enabled = canAdvance, weight = FontWeight.SemiBold, alpha = if (canAdvance) 1f else 0.3f,
                        ) { advance() }
                    }
                }
            }
        }
    }
}

@Composable
private fun QuestionGroup(
    title: String?, notes: List<String>, questions: List<SurveyQuestion>, answers: Map<String, String>,
    onSelect: (SurveyQuestion, SurveyChoice) -> Unit,
) {
    val ink = Ink
    Column(verticalArrangement = Arrangement.spacedBy(22.dp)) {
        if (title != null) Text(title, style = Fonts.almanac(19), color = ink.text)
        if (notes.isNotEmpty()) {
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                for (note in notes) Text(note, style = Fonts.system(13), color = ink.text.copy(alpha = 0.55f))
            }
        }
        for (question in questions) {
            val picked = SelfReportSurvey.values(answers[question.id])
            Column(verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text(question.text, style = Fonts.almanacBody(16, bold = true), color = ink.text, modifier = Modifier.padding(bottom = 4.dp))
                for (choice in question.choices) {
                    ChoiceRow(choice, selected = choice.value in picked, multiple = question.allowsMultiple) { onSelect(question, choice) }
                }
            }
        }
    }
}

// 문항 = 세리프 표제 + 라디오·괘선 리스트(책력 개방 조판). 측정 로직·저장 키는 무변경.
@Composable
private fun ChoiceRow(choice: SurveyChoice, selected: Boolean, multiple: Boolean, onClick: () -> Unit) {
    val ink = Ink
    Row(
        Modifier
            .fillMaxWidth()
            .semantics { this.selected = selected }
            .clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, onClick = onClick)
            .drawBehind { drawRect(ink.text.copy(alpha = 0.12f), topLeft = Offset(0f, size.height - 1f), size = Size(size.width, 1f)) }
            .padding(vertical = 11.dp),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        // 복수 문항은 라디오가 아니라 체크 — 하나만 고르는 줄로 읽히면 여러 개를 고를 생각을 못 한다
        if (multiple) CheckSquare(selected) else Radio(selected)
        Text(choice.label, style = Fonts.system(15, if (selected) FontWeight.SemiBold else FontWeight.Normal), color = ink.text.copy(alpha = if (selected) 1f else 0.75f))
    }
}

/** SF circle / circle.inset.filled 대용 */
@Composable
private fun Radio(selected: Boolean) {
    val ink = Ink
    val color = if (selected) ink.text else ink.text.copy(alpha = 0.3f)
    Box(
        Modifier.size(16.dp).drawBehind {
            drawCircle(color, style = androidx.compose.ui.graphics.drawscope.Stroke(width = 1.5.dp.toPx()), radius = size.minDimension / 2 - 1f)
            if (selected) drawCircle(color, radius = size.minDimension * 0.28f)
        },
    )
}

/** SF square / checkmark.square.fill 대용 */
@Composable
private fun CheckSquare(selected: Boolean) {
    val ink = Ink
    val color = if (selected) ink.text else ink.text.copy(alpha = 0.3f)
    val paper = ink.paper
    Box(
        Modifier.size(16.dp).drawBehind {
            val inset = 1.5.dp.toPx()
            val corner = androidx.compose.ui.geometry.CornerRadius(3.dp.toPx())
            if (selected) {
                drawRoundRect(color, topLeft = Offset(inset / 2, inset / 2), size = Size(size.width - inset, size.height - inset), cornerRadius = corner)
                val check = androidx.compose.ui.graphics.Path().apply {
                    moveTo(size.width * 0.27f, size.height * 0.52f)
                    lineTo(size.width * 0.44f, size.height * 0.69f)
                    lineTo(size.width * 0.74f, size.height * 0.35f)
                }
                drawPath(check, paper, style = androidx.compose.ui.graphics.drawscope.Stroke(width = 1.8.dp.toPx(), cap = androidx.compose.ui.graphics.StrokeCap.Round, join = androidx.compose.ui.graphics.StrokeJoin.Round))
            } else {
                drawRoundRect(color, topLeft = Offset(inset, inset), size = Size(size.width - inset * 2, size.height - inset * 2), cornerRadius = corner, style = androidx.compose.ui.graphics.drawscope.Stroke(width = inset))
            }
        },
    )
}

@Composable
private fun ActionText(title: String, enabled: Boolean, weight: FontWeight, alpha: Float, onClick: () -> Unit) {
    val ink = Ink
    Text(
        title, style = Fonts.system(17, weight), color = ink.text.copy(alpha = alpha),
        modifier = Modifier.clickable(interactionSource = remember { MutableInteractionSource() }, indication = null, enabled = enabled, onClick = onClick).padding(4.dp),
    )
}

// 진행도 명시(2026-08-08 베타 피드백 "진행도도 표시해줘")
@Composable
private fun ProgressHeader(step: Int) {
    val ink = Ink
    val total = SurveyLogic.TOTAL_STEPS
    val label = stringResource(R.string.survey_progress_a11y, step, total)
    Row(Modifier.fillMaxWidth().semantics { contentDescription = label }, verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
        Box(
            Modifier.weight(1f).height(4.dp).background(ink.text.copy(alpha = 0.12f), CircleShape).drawBehind {
                val w = maxOf(if (step > 0) 6.dp.toPx() else 0f, size.width * step / total)
                drawRoundRect(ink.text.copy(alpha = 0.65f), size = Size(w, size.height), cornerRadius = androidx.compose.ui.geometry.CornerRadius(size.height / 2))
            },
        )
        if (step > 0) Text(stringResource(R.string.survey_progress, step, total), style = Fonts.system(12).copy(fontFeatureSettings = "tnum"), color = ink.text.copy(alpha = 0.55f))
    }
}

@Composable
private fun Intro(prefilled: Boolean) {
    val ink = Ink
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(stringResource(R.string.survey_intro_title), style = Fonts.almanac(26), color = ink.text)
        Text(stringResource(R.string.survey_intro_body), style = Fonts.almanacBody(17), color = ink.text.copy(alpha = 0.75f))
        Text(stringResource(R.string.survey_intro_local), style = Fonts.system(13), color = ink.text.copy(alpha = 0.5f))
        if (prefilled) Text(stringResource(R.string.survey_intro_prefilled), style = Fonts.system(13), color = ink.text.copy(alpha = 0.5f))
    }
}

/** 증상 문항의 시간 앵커 — TempoCore `symptomAnchorLine`은 라벨을 문장에 합성하므로 번역 키가 아니다. 합성만 앱으로 가져온다. */
@Composable
private fun symptomAnchorTitle(p2: String?): String {
    // 복수 응답(2026-09-09)이면 고른 것을 전부 나열한다 — tempocore symptomAnchorLine과 같은 규칙.
    val labels = SelfReportSurvey.values(p2)
        .filter { it !in SelfReportSurvey.exclusivePhaseValues }
        .mapNotNull { value -> SelfReportSurvey.phaseChoices.firstOrNull { it.value == value }?.label }
    if (labels.isEmpty()) return stringResource(R.string.survey_anchor_fallback)
    return stringResource(R.string.survey_anchor_format, labels.joinToString(" · "))
}
